packages/teleterm/src/ui/ClusterLogin/useClusterLogin.ts

```diff
@@ -85,7 +85,10 @@ export default function useClusterLogin(props: Props) {
 
     if (loginAttempt.status === 'success') {
       onClose();
-      serviceDocs.open(clusterUri);
+      if (!customOnSuccess) {
+        serviceDocs.open(clusterUri);
+      }
+      customOnSuccess?.();
```

Comment:
Passing this callback through many components/actions seem too fragile. Lets try to add a 

```ts
    postLoginActions: new Map<{clusterUri, cb: ()=> void > ()
```

to the `clustersService` and allow to register a single post-action callback per cluster. The `commandLauncher` would be the place to initialize these post login actions, something like below:

```ts
  'cluster-login': {
    displayName: '',
    description: '',
    run(ctx: IAppContext, args: { clusterUri?: string, done }) {
      if(done){
        ctx.clustersService.setPostLoginAction(clusterUri, done)
      }else{
        // default action
        ctx.clustersService.setPostLoginAction(clusterUri, () => {
          serviceDocs.open(clusterUri);
        })
      }
      
      if (!args.clusterUri) {
        ctx.modalsService.openAddClusterDialog();
      } else {
        ctx.modalsService.openLoginDialog(args.clusterUri);
      }
    },
  },
 ```

Reply:
Hmm it won't work for `openAddClusterDialog()` case - after adding a cluster it opens login dialog, but as we didn't  have a `clusterUri` we couldn't set an action for it (event default).

I can agree that passing down props has downsides, but I'm afraid that proposed solution can increase complexity (some of the props are passed directly to the modal, some properties are stored globally; where should we add new properties?). I'd like to avoid a global state and try to keep the modal as _clean_ as possible.

What would you say for renaming `customOnSuccess` to `onSuccess` and moving `serviceDocs.open(clusterUri);` from `useClusterLogin` to `commandLauncher`? By doing this we will remove that high-level logic from lower-level components.

Reply:
> Hmm it won't work for openAddClusterDialog() case - after adding a cluster it opens login dialog

Oh yeah, that was for illustration purposes to give an idea, `openAddClusterDialog()` needs to be removed from that command (or be a separate command instead). 

> I can agree that passing down props has downsides, but I'm afraid that proposed solution can increase complexity (some of the props are passed directly to the modal, some properties are stored globally; where should we add new properties?).

My thinking was to decouple `dialogs` from execution of post login actions by moving `postLoginActions` to the `clusterService`. This way it is possible to subscribe to these events without dependency on the dialog itself (via cb). The logout actions would work similarly.

> (some of the props are passed directly to the modal, some properties are stored globally; where should we add new properties?)

`dialog` has own properties, `clusterServices`, and `commandLauncher` commands theirs. `Command` encapsulates the high level relationships between these. To open a login dialog, one should execute a command with arguments. The command then opens a dialog and interacts with other services to carry out its execution.  

> What would you say for renaming customOnSuccess to onSuccess and moving serviceDocs.open(clusterUri); from useClusterLogin to commandLauncher

Yeah, that was the main idea - to decouple dialog from `docsService` or any "on-login" event handlers. If `postLoginActions` idea seems raw at the moment, lets postpone the discussion and proceed with `onSuccess` callback per your suggestion.

Reply:
I understand your points, some questions:
1. How would we handle `openAddClusterDialog()` to preserve current behavior (opening add dialog -> opening login dialog -> opening document)? It's used in `ExpanderCluster` for clicking an add icon. We can't set a callback per `clusterUri` as we don't know it yet.
2. Should an action be cleared after running it? For example, if someone needs to just run `clusterService.login()` directly, with no side effects, should he run `ctx.clustersService.setPostLoginAction(clusterUri, null)` first?